api/run.go

```diff
@@ -0,0 +1,111 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package api allows go projects to use func-e as a library.
+package api
+
+import (
+	"context"
+	"io"
+	"os"
+	"runtime"
+
+	"github.com/tetratelabs/func-e/internal/cmd"
+	"github.com/tetratelabs/func-e/internal/globals"
+	"github.com/tetratelabs/func-e/internal/version"
+)
+
+// HomeDir is an absolute path which most importantly contains "versions"
+// installed from EnvoyVersionsURL. Defaults to "${HOME}/.func-e"
+func HomeDir(homeDir string) RunOption {
+	return func(o *runOpts) {
+		o.homeDir = homeDir
+	}
+}
+
+// EnvoyVersionsURL is the path to the envoy-versions.json.
+// Defaults to "https://archive.tetratelabs.io/envoy/envoy-versions.json"
+func EnvoyVersionsURL(envoyVersionsURL string) RunOption {
+	return func(o *runOpts) {
+		o.envoyVersionsURL = envoyVersionsURL
+	}
+}
+
+// EnvoyVersion overrides the version of Envoy to run. Defaults to the
+// contents of "$HomeDir/versions/version".
+//
+// When that file is missing, it is generated from ".latestVersion" from the
+// EnvoyVersionsURL. Its value can be in full version major.minor.patch format,
+// e.g. 1.18.1 or without patch component, major.minor, e.g. 1.18.
+func EnvoyVersion(envoyVersion string) RunOption {
+	return func(o *runOpts) {
+		o.envoyVersion = envoyVersion
+	}
+}
+
+// Out is where status messages are written. Defaults to os.Stdout
+func Out(out io.Writer) RunOption {
+	return func(o *runOpts) {
+		o.out = out
+	}
+}
+
+// RunOption is configuration for Run.
+type RunOption func(*runOpts)
+
+type runOpts struct {
+	homeDir          string
+	envoyVersion     string
+	envoyVersionsURL string
+	out              io.Writer
+}
+
+// Run downloads Envoy and runs it as a process with the arguments
+// passed to it. Use RunOption for configuration options.
+func Run(ctx context.Context, args []string, options ...RunOption) error {
+	ro := &runOpts{
+		homeDir:          globals.DefaultHomeDir,
+		envoyVersion:     "", // default to lookup
+		envoyVersionsURL: globals.DefaultEnvoyVersionsURL,
+		out:              os.Stdout,
+	}
+	for _, option := range options {
+		option(ro)
+	}
+
+	o := globals.GlobalOpts{
+		HomeDir:          ro.homeDir,
+		EnvoyVersion:     version.PatchVersion(ro.envoyVersion),
+		EnvoyVersionsURL: ro.envoyVersion,
+		Out:              ro.out,
+	}
+
+	funcECmd := cmd.NewApp(&o)
+
+	funcERunArgs := []string{"func-e", "--platform", runtime.GOOS + "/" + runtime.GOARCH, "run"}
+	funcERunArgs = append(funcERunArgs, args...)
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- funcECmd.RunContext(ctx, funcERunArgs)
+	}()
+
+	// Wait for run to exit or an explicit stop.
+	select {
+	case <-ctx.Done():
+		return nil
```

Comment:
shouldn't we kill the envoy process in this case? (not sure if there's such function in func-e now)

Reply:
isn't this done in the called func libs https://github.com/tetratelabs/func-e/blob/82577c96a87fc3a5c5d7639add588cadb3b5fe9c/internal/envoy/run.go#L78 ?

Reply:
ok cool but it seems to me that the code block only reaches when the OS signal was sent. Is there any way to verify that it works as intended? e.g. passing the fake config that runs indefinitely in tests?

Reply:
can you point me to some tests in the code base that verify this, that I can peruse here ?